feat: fastest path endpoint

JackSMunn opened the pull request:
This pull request introduces several new configurations and improvements to the testing and linting workflows, as well as updates to the application code to enhance functionality and maintainability. The most important changes include the addition of coverage configuration, new GitHub workflows for linting and testing, and significant updates to the `gateway` route and its corresponding tests.

### Configuration and Workflow Improvements:

* [`.coveragerc`](diffhunk://#diff-834e9b406d74791ffbafaeec9cc894082cea9739bc347b6ff9f72312c92ebc79R1-R13): Added coverage configuration to omit specific files and show missing lines in the report.
* [`.github/workflows/lint.yml`](diffhunk://#diff-107e910e9f2ebfb9a741fa10b2aa7100cc1fc4f5f3aca2dfe78b905cbd73c0d2R1-R28): Introduced a new GitHub Actions workflow for running Ruff linter on push and pull request events.
* [`.github/workflows/test.yml`](diffhunk://#diff-faff1af3d8ff408964a57b2e475f69a6b7c7b71c9978cccc8f471798caac2c88R1-R32): Added a GitHub Actions workflow to run PyTest on push and pull request events.
* [`pytest.ini`](diffhunk://#diff-fb6a686182f16eb54af3c628f38593f347f68aba31de903983023c560288d7a1R1-R2): Configured PyTest to use importlib mode and added coverage options.
* [`ruff.toml`](diffhunk://#diff-26f436cb29eecc7bb4f9066747b43df8da81ac808955ce95c109f5d7aa778989R1-R6): Set up Ruff linter configuration with specific formatting rules.

### Application Code Enhancements:

* [`app/main.py`](diffhunk://#diff-990f7e4140fab1ad82afc787505090b64d4024e63a891405cd9085e7e6b249ddR5-R9): Added import statements for `sys`, `os`, and `uvicorn`, and updated the uvicorn run command for better clarity and consistency. [[1]](diffhunk://#diff-990f7e4140fab1ad82afc787505090b64d4024e63a891405cd9085e7e6b249ddR5-R9) [[2]](diffhunk://#diff-990f7e4140fab1ad82afc787505090b64d4024e63a891405cd9085e7e6b249ddL23-L25)
* [`app/routes/gateway.py`](diffhunk://#diff-ec7995512a23a58c4b4bef364f8161d8ec07baac3ac186250788cbbd587584c0L1-R82): Refactored the gateway route to include a new endpoint for pathfinding, improved error handling, and added comments for clarity.

### Testing Improvements:

* [`app/test/test_forwarder.py`](diffhunk://#diff-db0aef0a3a85d0ff2a6788258d2d3937c9187a48852cb904db304295bfb660beR1-R62): Added tests for the `forward_request` utility function, including success and error scenarios.
* [`app/test/test_gateway.py`](diffhunk://#diff-02566e96156a33593603ed806a3ef67d5d3a2f1fad163347590c0cc160d7c764R1-R122): Introduced comprehensive tests for the gateway route, including tests for guard clauses, error handling, and successful flow.

JackSMunn committed:
feat: fastest_path forwarding

- changed gateway.py endpoint that handles "fastest_path" from frontend
- added init files
- changed main.py

JackSMunn committed:
updated: gateway.py to correct paths

- changed gateway.py to allow for the correct paths /fastest-path, and pathfinding/fastest-path
- consider homogenizing names for the JSON packages henceforth.

JackSMunn committed:
feat: actions

- added linting with ruff
- added testing with pytest

JackSMunn committed:
test: unittesting for gateway and forwarder

- added coveragerc config
- outcommented some imports out for linting
- added unit tests for forwarder
- added unit tests for gateway with focus on the guards
- added pytest config
- added ruff config

JackSMunn committed:
fix: pytest.ini

JackSMunn committed:
update: requirements

Mafusn commented on app/config.py:
```diff
@@ -1,4 +1,4 @@
-from fastapi.middleware.cors import CORSMiddleware
+#from fastapi.middleware.cors import CORSMiddleware
```
Should this line be deleted?

Mafusn reviewed (changes requested):
Very small comment, and waiting for route/gateway.py functionality to be moved into a controller instead

JackSMunn committed:
update: split gateway.py

- split the gateway file into controller, routes & schema
- changed main to new import
- deleted gateway.py
- added api_routes.py to congregate the logic
- added init files
- added pathfinding_schema.py that defines the pathfindingrequest schema
- added pathfinding_service.py that contains the logic for the fastest_path endpoint
- changed tests to reflect new file structure

Mafusn reviewed (approved):
LGTM 🧄

JackSMunn committed:
fix: deleted unused import in config + update readme

Mafusn reviewed (approved):
LVGTM 🍖

MortenMunk reviewed (approved):
lgtm :denmark: :+1:  :1st_place_medal:
